File: app/src/main/java/com/example/shapesapp/views/StatsActivity.java
package com.example.shapesapp.views;

import android.os.Bundle;

import android.util.Log;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.shapesapp.R;
import com.example.shapesapp.adapter.RecyclerViewEmptyObserver;
import com.example.shapesapp.models.Shape;
import com.example.shapesapp.presenter.StatsPresenter;

import java.util.HashMap;


public class StatsActivity extends AppCompatActivity {

    private RecyclerView mRecyclerView;
    private StatsAdapter mAdapter;
    private RecyclerView.LayoutManager mLayoutManager;
    private TextView statsEmptyView;
    StatsPresenter statsPresenter;
    private HashMap<Shape.Type, Integer> myDataset;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_stats);
//        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
//        setSupportActionBar(toolbar);

        mRecyclerView = (RecyclerView) findViewById(R.id.stats_recycler_view);
        statsEmptyView = (TextView) findViewById(R.id.emptyView);
        statsPresenter = new StatsPresenter();
        setupStatsListView();
    }

    private void setupStatsListView() {

        mRecyclerView.setHasFixedSize(true);

        mLayoutManager = new LinearLayoutManager(this);
        mRecyclerView.setLayoutManager(mLayoutManager);

        myDataset = (HashMap<Shape.Type, Integer>) statsPresenter.getCountByGroup();
        statsEmptyView.setVisibility(View.GONE);
        mAdapter = new StatsAdapter(myDataset, this, onClickDelete);
        mRecyclerView.setAdapter(mAdapter);
        mAdapter.registerAdapterDataObserver(new RecyclerViewEmptyObserver(mRecyclerView, statsEmptyView));
    }

    StatsAdapter.OnItemClicked onClickDelete = new StatsAdapter.OnItemClicked() {
        @Override
        public void onItemClick(int position) {
            Shape.Type type = (Shape.Type) myDataset.keySet().toArray()[position];
            Log.d("canvas1234", "   on item click : type " + type);
            statsPresenter.deleteAllByShape(type);
            Toast.makeText(StatsActivity.this, " All "+type+"s deleted. ", Toast.LENGTH_SHORT).show();
            refreshData();
        }
    };

    private void refreshData() {
        myDataset = (HashMap<Shape.Type, Integer>) statsPresenter.getCountByGroup();
        mAdapter.setmDataSet(myDataset);
        mAdapter.notifyDataSetChanged();
    }

}

File: app/src/main/java/com/example/shapesapp/views/StatsActivity.kt
package com.example.shapesapp.views

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.shapesapp.R
import com.example.shapesapp.adapter.RecyclerViewEmptyObserver
import com.example.shapesapp.adapter.StatsAdapter
import com.example.shapesapp.databinding.ActivityStatsBinding
import com.example.shapesapp.models.Shape
import com.example.shapesapp.presenter.StatsPresenter

class StatsActivity : AppCompatActivity() {
    private var mRecyclerView: RecyclerView? = null
    private var mAdapter: StatsAdapter? = null
    private var mLayoutManager: RecyclerView.LayoutManager? = null
    private var statsEmptyView: TextView? = null
    var statsPresenter: StatsPresenter? = null
    private var myDataset: HashMap<Shape.Type, Int>? = null
    private lateinit var binding: ActivityStatsBinding
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_stats)
        mRecyclerView = binding.statsRecyclerView
        statsEmptyView = binding.emptyView
        statsPresenter = StatsPresenter()
        setupStatsListView()
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    }

    private fun setupStatsListView() {
        mRecyclerView?.setHasFixedSize(true)
        mLayoutManager = LinearLayoutManager(this)
        mRecyclerView?.layoutManager = mLayoutManager
        myDataset = statsPresenter?.countByGroup as HashMap<Shape.Type, Int>
        statsEmptyView?.visibility = View.GONE
        mAdapter = StatsAdapter(myDataset)
        mRecyclerView?.adapter = mAdapter
        mAdapter?.registerAdapterDataObserver(
            RecyclerViewEmptyObserver(
                mRecyclerView!!,
                statsEmptyView
            )
        )
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}